File: sdks/nodejs/src/context.ts
import type { SessionInfo, UserProfile } from './types';
import type { ZroApp } from './app';

/** Context passed to every command and event handler. */
export class AppContext {
    readonly session: SessionInfo;
    readonly instanceId: string | null;
    readonly slug: string;
    readonly dataDir: string;
    private _app: ZroApp;

    constructor(
        session: SessionInfo,
        instanceId: string | null,
        slug: string,
        dataDir: string,
        app: ZroApp,
    ) {
        this.session = session;
        this.instanceId = instanceId;
        this.slug = slug;
        this.dataDir = dataDir;
        this._app = app;
    }

    /** Emit an event to a specific instance. */
    emitTo(instanceId: string, event: string, data: any): void {
        this._app._emit({ type: 'instance', instance_id: instanceId }, event, data);
    }

    /** Broadcast an event to all instances of this app. */
    emit(event: string, data: any): void {
        this._app._emit({ type: 'broadcast' }, event, data);
    }

    /** Emit an event to all apps within the current user session. */
    emitToSession(event: string, data: any): void {
        this._app._emit({ type: 'session', session_id: this.session.session_id }, event, data);
    }

    /** Emit a system-wide event to every connected client. */
    emitSystem(event: string, data: any): void {
        this._app._emit({ type: 'system' }, event, data);
    }

    /** Get the user profile (if available). */
    get profile(): UserProfile | undefined {
        return this.session.profile;
    }

    /** Get the current username. */
    get username(): string {
        return this.session.username;
    }

    /** Get the current user's role. */
    get role(): string {
        return this.session.role;
    }

    /** Get the current user's groups. */
    get groups(): string[] {
        return this.session.groups;
    }

    /** Access registered state by key. */
    state(key: string): any {
        return this._app._getState(key);
    }
}

File: sdks/rust/src/context.rs
use std::path::PathBuf;
use std::sync::Arc;

use tokio::sync::Mutex;

use zro_protocol::messages::{write_message, EventEmitPayload, EventTarget, IpcMessage};
use zro_protocol::types::SessionInfo;

/// Context provided to every command handler, event handler, and lifecycle hook.
///
/// Provides access to session info, the IPC writer for emitting events, and app metadata.
///
/// - In WS invoke / WS event handlers: `instance_id` is `Some(...)`.
/// - In HTTP API handlers: `instance_id` is `None`.
#[derive(Clone)]
pub struct AppContext {
    pub session: SessionInfo,
    pub instance_id: Option<String>,
    pub slug: String,
    pub data_dir: PathBuf,
    writer: Arc<Mutex<tokio::net::unix::OwnedWriteHalf>>,
}

impl AppContext {
    pub(crate) fn new(
        session: SessionInfo,
        instance_id: Option<String>,
        slug: String,
        data_dir: PathBuf,
        writer: Arc<Mutex<tokio::net::unix::OwnedWriteHalf>>,
    ) -> Self {
        Self {
            session,
            instance_id,
            slug,
            data_dir,
            writer,
        }
    }

    /// Emit an event to a specific client instance.
    pub async fn emit_to(
        &self,
        instance_id: &str,
        event: &str,
        payload: serde_json::Value,
    ) -> Result<(), crate::app::ZroSdkError> {
        let emit = EventEmitPayload {
            event: event.to_string(),
            payload,
            target: EventTarget::Instance {
                instance_id: instance_id.to_string(),
            },
        };
        let msg = IpcMessage::new("EventEmit", serde_json::to_value(emit)?);
        let mut w = self.writer.lock().await;
        write_message(&mut *w, &msg).await?;
        Ok(())
    }

    /// Broadcast an event to all connected clients of this app.
    pub async fn emit(
        &self,
        event: &str,
        payload: serde_json::Value,
    ) -> Result<(), crate::app::ZroSdkError> {
        let emit = EventEmitPayload {
            event: event.to_string(),
            payload,
            target: EventTarget::Broadcast,
        };
        let msg = IpcMessage::new("EventEmit", serde_json::to_value(emit)?);
        let mut w = self.writer.lock().await;
        write_message(&mut *w, &msg).await?;
        Ok(())
    }
}

impl std::fmt::Debug for AppContext {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("AppContext")
            .field("session", &self.session)
            .field("instance_id", &self.instance_id)
            .field("slug", &self.slug)
            .finish()
    }
}
